### PR TITLE
Add-DbaPushSubscription

### PR DESCRIPTION
**Pull request by garethnewman**
## Type of Change

 - [X] New feature (non-breaking change, adds functionality)

### Purpose
Working on a new command, will need help to bring it up to dbatools standards please, go easy this is new to me 🌚 

Tested and working on my local sql server 2017, not sure what the next steps are to get it up to dbatools standards. help :/

### CODE REVIEW

**Comment by potatoqualitee** [comment]
yay! thank you 🎉  we are in desperate need of replication commands. i'll take a look momentarily.

File: functions/Add-DbaPushSubscription.ps1 [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        $subscriberName,
+        $publicationName,
+        $pubDatabaseName,
+        $subDatabaseName,
+        $subSyncType
+    )
+
+    # create the server connection object
+    try {
+        $srv = New-Object Microsoft.SqlServer.Management.Smo.Server $publisherName
```
Comment (potatoqualitee) [inline functions/Add-DbaPushSubscription.ps1]: Thanks so much for the command! Please use the templates at [dbatools.io/templates](https://dbatools.io/templates). In particular, this one: https://github.com/sqlcollaborative/dbatools-templates/blob/master/Get-DbaSmoObject.ps1

Please add Rep this - `Add-DbaRepTherest....` but what do you think about `Add-DbaRepSubscription -Type Push` so you can support all the diff subscription types from one command? (asking, not commanding, im don't know enough about replication to say, really)

**Comment by wsmelton** [comment]
I would see this as being called `New-Dba...` and not `Add`. On the type I think it will be easier to handle testing if we have separate commands. There are actually 4 different subscriptions you can have: Push, Pull and then merge replication you add two additional client and server.

But up to y'all.

**Comment by potatoqualitee** [comment]
(all new commands will have to be published to the prerelease due to all of the changes we've made in psd1 and psm1)

**Comment by potatoqualitee** [comment]
it says outdated but i cant see the changes. can anyone else?

**Comment by nvarscar** [comment]
> it says outdated but i cant see the changes. can anyone else?

Do you see the comments at least? There were no specific code changes provided, just comments.

**Comment by potatoqualitee** [comment]
i do 🤔 maybe it'll go away with next commit

**Comment by nvarscar** [comment]
My guess is that they became outdated because the branch was changed. But ¯\\\_(ツ)\_/¯

**Comment by potatoqualitee** [comment]
@garethnewman any chance you can get some time to work on this? we'd like to close PRs ASAP. I'm going to be doing mass changes to formatting coming up shortly (shawn, just running every ps1 against invoke-dbatools formatter)

**Comment by garethnewman** [comment]
Yes, Sorry I missed the comments somehow, I see everything now. I'll get on it this weekend.

File: functions/Add-DbaRepPushSubscription.ps1 [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    } catch {
+        Stop-Function -Message "Error occurred while establishing connection to $publisherName" -Category ConnectionError -ErrorRecord $_ -Target $publisherName -Continue
+    }
+
+
+    # create the PULL replication subscription object
+    try{
+        $ts = New-Object Microsoft.SqlServer.Replication.TransSubscription
+    } catch {
+        Stop-Function -Message "Error occurred creating the replication object" -Continue
```
[the remark this answers is not in the document]
Comment (garethnewman) [reply functions/Add-DbaRepPushSubscription.ps1]: done

File: functions/Add-DbaRepPushSubscription.ps1 [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    #>
+    Param(
+        [Parameter(Mandatory)]
+        $publisherName,
+        $subscriberName,
+        $publicationName,
+        $pubDatabaseName,
+        $subDatabaseName,
+        $subSyncType
```
[the remark this answers is not in the document]
Comment (garethnewman) [reply functions/Add-DbaRepPushSubscription.ps1]: done

File: functions/Add-DbaRepPushSubscription.ps1 [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    .EXAMPLE
+        PS C:\> Add-DbaRepPushSubscription -publisherName "server1\dev2016" -subscriberName "server2\dev2016" -publicationName "avworkspub1" -pubDatabaseName "AdventureWorks2012" -subDatabaseName "replicatedAdventureWorks2012" -subSyncType "none"
+
+        Creates a new subscription to "avworkspub1" on server1\dev2016 to server2\dev2016 with a subscription sync type of none.
+        The published database is AdventureWorks2012 on server1\dev2016, the subscriber database is replicatedAdventureWorks2012 on server2\dev2016.
+
+    #>
+    Param(
+        [Parameter(Mandatory)]
+        $publisherName,
```
[the remark this answers is not in the document]
Comment (garethnewman) [reply functions/Add-DbaRepPushSubscription.ps1]: done

**Comment by garethnewman** [comment]
I have done all the easy bits, but I realise now I need some extra stuff to handle the other sync types. I'll close the PR and return when it's ready so it's not sitting there for a few weeks.

**Pull request closed without merging**